### PR TITLE
Have MortarData store DataVectors rather than std::vector

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

This is a step towards #4580. Because we will eventually be using charm messages to send the boundary data and we want to avoid copies when we are on the same node, the MortarData needs to be able to store non-owning vectors and std::vector doesn't support this. That's why we make this switch to using DataVectors.

While making this change, I also had to change everything that uses the MortarData. This included ComputeTimeDerivaive, ApplyBoundaryCorrections, and all the subcell code because it previously expected the neighbor data to be a std::vector.

Basically all of the changes are in the last commit.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/Actions/ComputeTimeDerivativeImpl.tpp [hunk 1]
```diff
@@ -1750,17 +1752,26 @@ void test_impl(const Spectral::Quadrature quadrature,
...
-            boundary_data_on_mortar.data() + boundary_data_on_mortar.size()};
+
+        DataVector expected_data{};
+        if (Spectral::needs_projection(face_mesh, mortar_mesh, mortar_size)) {
+          expected_data = DataVector{
+              mortar_mesh.number_of_grid_points() *
+              Variables<mortar_tags_list>::number_of_independent_components};
+          Variables<mortar_tags_list> projected_packaged_data{
+              expected_data.data(), expected_data.size()};
+          const auto blah = ::dg::project_to_mortar(packaged_data, face_mesh,
```
Comment (nilsdeppe) [inline tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/Actions/ComputeTimeDerivativeImpl.tpp]: maybe rename `blah` to something a bit clearer? I think you did this in `src` already

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
LGTM, one suggestion.  Thank you so much for doing this! Please squash immediately because I don't want to hold this up very long, given the number of files it changes :)

File: tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/Actions/ComputeTimeDerivativeImpl.tpp (hunk 1 above)
Comment (knelli2) [reply tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/Actions/ComputeTimeDerivativeImpl.tpp]: Oops! Yeah forgot to change that. I just copied what I did in `InternalMortarDataImpl.hpp` (also copied the comment that went along with it for clarity)

**Comment by knelli2** [comment]
Also rebased on develop

**Review by nilsdeppe — APPROVED** [review]
Thank you so much for doing this! :D Nice work!
